dataplane/api-server/src/lib.rs

```diff
@@ -5,33 +5,128 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
 */
 
 pub mod backends;
+pub mod config;
 pub mod netutils;
 pub mod server;
 
-use std::net::{Ipv4Addr, SocketAddrV4};
+use std::{
+    fs,
+    net::{Ipv4Addr, SocketAddrV4},
+};
 
-use anyhow::Error;
+use anyhow::{Context, Result};
 use aya::maps::{HashMap, MapData};
-use tonic::transport::Server;
+use log::info;
+use tonic::transport::{Certificate, Identity, Server, ServerTlsConfig};
 
 use backends::backends_server::BackendsServer;
 use common::{BackendKey, BackendList, ClientKey, LoadBalancerMapping};
+use config::TLSConfig;
 
 pub async fn start(
     addr: Ipv4Addr,
     port: u16,
     backends_map: HashMap<MapData, BackendKey, BackendList>,
     gateway_indexes_map: HashMap<MapData, BackendKey, u16>,
     tcp_conns_map: HashMap<MapData, ClientKey, LoadBalancerMapping>,
-) -> Result<(), Error> {
-    let (_, health_service) = tonic_health::server::health_reporter();
-
-    let server = server::BackendService::new(backends_map, gateway_indexes_map, tcp_conns_map);
-    // TODO: mTLS https://github.com/Kong/blixt/issues/50
-    Server::builder()
-        .add_service(health_service)
-        .add_service(BackendsServer::new(server))
-        .serve(SocketAddrV4::new(addr, port).into())
-        .await?;
+    tls_config: Option<TLSConfig>,
+) -> Result<()> {
+    // Tonic itself doesn't provide a built-in mechanism for selectively
+    // applying TLS based on routes, as TLS configuration is tied to the
+    // entire server and managed at the transport layer, not at the
+    // application layer where routes are defined.
+    //
+    // Solution: separate gRPC services
+    //
+    // Public server without mTLS for healthchecks
+    let healthchecks = tokio::spawn(async move {
+        let (_, health_service) = tonic_health::server::health_reporter();
+        let mut server_builder = Server::builder();
+        server_builder
+            .add_service(health_service)
+            .serve(SocketAddrV4::new(addr, port + 1).into())
+            .await
+            .unwrap();
+    });
+
+    // Secure server with (optional) mTLS
+    let backends = tokio::spawn(async move {
+        let server = server::BackendService::new(backends_map, gateway_indexes_map, tcp_conns_map);
+        let mut server_builder = Server::builder();
+        server_builder = setup_tls(server_builder, &tls_config).unwrap();
+        server_builder
+            .add_service(BackendsServer::new(server))
+            .serve(SocketAddrV4::new(addr, port).into())
+            .await
+            .unwrap();
+    });
+
+    tokio::try_join!(healthchecks, backends)?;
+
     Ok(())
 }
+
+pub fn setup_tls(mut builder: Server, tls_config: &Option<TLSConfig>) -> Result<Server> {
+    // TLS implementation drawn from Tonic examples.
+    // See: https://github.com/hyperium/tonic/blob/master/examples/src/tls_client_auth/server.rs
```

Comment:
I always appreciate when we link to source materials and relevant external documentation, ty :+1: